### PR TITLE
Transformer for Concat

### PR DESCRIPTION
**Pull request by Zruty0**
Proper conversion of ConcatTransform

### CODE REVIEW

**Comment by Ivanidzo4ka** [comment]
>                     viewTrain);

Don't you want to get rid of calls to Arguments class?

---
Refers to: src/Microsoft.ML/Runtime/EntryPoints/FeatureCombiner.cs:79 in 0b239ea. [](commit_id = 0b239eab04e649022b56a3e69746064e3ed87bca, deletion_comment = False)

File: src/Microsoft.ML.Data/DataView/RowToRowMapperTransform.cs [hunk 1]
```diff
@@ -134,15 +134,15 @@ public sealed class RowToRowMapperTransform : RowToRowTransformBase, IRowToRowMa
     {
         private sealed class Bindings : ColumnBindingsBase
         {
-            private readonly RowToRowMapperTransform _parent;
+            private readonly IRowMapper _mapper;
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/DataView/RowToRowMapperTransform.cs]: >IRowMapper [](start = 29, length = 10)

Would this stop double mapper creation which we have right now for each RowToRowMapperTransform instantiation?

File: src/Microsoft.ML.Data/Transforms/ConcatEstimator.cs [hunk 2]
```diff
@@ -123,90 +112,6 @@ public SchemaShape GetOutputSchema(SchemaShape inputSchema)
         }
     }
 
-    // REVIEW: Note that the presence of this thing is a temporary measure only.
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/Transforms/ConcatEstimator.cs]: >// REVIEW: Note that the presence of this thing is a temporary measure only. [](start = 4, length = 76)

People spend time writing this code, and you deleting it without any remorse!

File: src/Microsoft.ML.Data/Transforms/ConcatTransform.cs [hunk 3]
```diff
@@ -18,19 +18,33 @@
 using Microsoft.ML.Runtime.Model.Onnx;
 using Microsoft.ML.Runtime.Model.Pfa;
 using Newtonsoft.Json.Linq;
+using Microsoft.ML.Core.Data;
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/Transforms/ConcatTransform.cs]: mildly shock

File: src/Microsoft.ML.Data/Transforms/ConcatTransform.cs [hunk 4]
```diff
@@ -113,6 +127,218 @@ public sealed class TaggedArguments
...
+            }
+
+            var result = new ColumnInfo[n];
+            for (int i = 0; i < n; i++)
+                result[i] = new ColumnInfo(names[i],
+                    inputs[i].Zip(aliases[i], (name, alias) => (name, alias)));
+            return result;
+        }
+
         private sealed class Bindings : ManyToOneColumnBindingsBase
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/Transforms/ConcatTransform.cs]: >Bindings [](start = 29, length = 8)

You probably forgot to delete this peace of code.

File: src/Microsoft.ML.Data/Transforms/ConcatTransform.cs [hunk 5]
```diff
@@ -113,6 +127,218 @@ public sealed class TaggedArguments
...
+                return inputNames.Select(name => (name, (string)null));
+            }
+
+            public ColumnInfo(string outputName, IEnumerable<(string name, string alias)> inputs)
+            {
+                Contracts.CheckNonEmpty(outputName, nameof(outputName));
+                Contracts.CheckValue(inputs, nameof(inputs));
+                Contracts.CheckParam(inputs.Any(), nameof(inputs), "Can not be empty");
+
+                foreach (var (name, alias) in inputs)
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/Transforms/ConcatTransform.cs]: >alias [](start = 36, length = 5)

Can we have a test in which we use these aliases, or maybe some kind of ///summary on top of this constructor with description of aliases.
Right now it's mystery for me, and any potential user.

**Comment by TomFinley** [comment]
>                     viewTrain);

*Internally* it is fine for now. Externally, yes I agree.

---
In reply to: [420842898](https://github.com/dotnet/machinelearning/pull/896#issuecomment-420842898) [](ancestors = 420842898)

---
Refers to: src/Microsoft.ML/Runtime/EntryPoints/FeatureCombiner.cs:79 in 0b239ea. [](commit_id = 0b239eab04e649022b56a3e69746064e3ed87bca, deletion_comment = False)

File: src/Microsoft.ML.Data/Transforms/ConcatTransform.cs (hunk 3 above)
Comment (TomFinley) [reply src/Microsoft.ML.Data/Transforms/ConcatTransform.cs]: Indeed Pete, you must learn to sort. :D :D :D

---
In reply to: [217230842](https://github.com/dotnet/machinelearning/pull/896#discussion_r217230842) [](ancestors = 217230842)

File: src/Microsoft.ML.Data/DataView/RowToRowMapperTransform.cs (hunk 1 above)
Comment (Zruty0) [reply src/Microsoft.ML.Data/DataView/RowToRowMapperTransform.cs]: I think yes

---
In reply to: [217230609](https://github.com/dotnet/machinelearning/pull/896#discussion_r217230609) [](ancestors = 217230609)

File: src/Microsoft.ML.Data/Transforms/ConcatTransform.cs (hunk 4 above)
Comment (Zruty0) [reply src/Microsoft.ML.Data/Transforms/ConcatTransform.cs]: correct

---
In reply to: [217231847](https://github.com/dotnet/machinelearning/pull/896#discussion_r217231847) [](ancestors = 217231847)

**Comment by Zruty0** [comment]
>                     viewTrain);

yes


---
In reply to: [420868137](https://github.com/dotnet/machinelearning/pull/896#issuecomment-420868137) [](ancestors = 420868137,420842898)

---
Refers to: src/Microsoft.ML/Runtime/EntryPoints/FeatureCombiner.cs:79 in 0b239ea. [](commit_id = 0b239eab04e649022b56a3e69746064e3ed87bca, deletion_comment = False)

File: src/Microsoft.ML.Data/Transforms/ConcatEstimator.cs (hunk 2 above)
Comment (Zruty0) [reply src/Microsoft.ML.Data/Transforms/ConcatEstimator.cs]: Extra credit for removing Tom's code!

---
In reply to: [217230777](https://github.com/dotnet/machinelearning/pull/896#discussion_r217230777) [](ancestors = 217230777)

Comment (TomFinley) [reply src/Microsoft.ML.Data/Transforms/ConcatEstimator.cs]: If a code falls in the forest...

---
In reply to: [217551182](https://github.com/dotnet/machinelearning/pull/896#discussion_r217551182) [](ancestors = 217551182,217230777)

**Pull request merged**